pkg/controller/config/dns.go

```diff
@@ -97,50 +147,83 @@ func (h *configHandler) Handle(req router.Request, resp router.Response) error {
 		sec.Annotations[labels.AcornDNSState] = state
 		sec.Data = map[string][]byte{"domain": []byte(domain), "token": []byte(token)}
 		if err := req.Client.Update(req.Ctx, sec); err != nil {
-			return err
+			return "", err
 		}
 	}
 
 	if !strings.EqualFold(*cfg.LetsEncrypt, "disabled") {
 		if err := tls.ProvisionWildcardCert(req, resp, domain, token); err != nil {
-			return err
+			return "", err
 		}
 	}
 
-	return nil
+	return domain, nil
 }
 
-// purgeRecordsIfDisabling checks if we are transitioning AcorDNS from an enabled state to a disabled state and if so calls the
-// acorn DNS service to purge all records for the domain. It is expected that string variable returned by this function
-// will be set as the labels.AcornDNSState annotation on the acorn-dns secret
-func purgeRecordsIfDisabling(req router.Request, domain string, cfg *apiv1.Config, dnsSecret *corev1.Secret,
-	token string, dnsClient dns.Client) (string, error) {
-	var state string
-	if domain == "" {
-		state = *cfg.AcornDNS
-	} else {
-		// The config object includes the acorn domain in the list of ClusterDomains if we are in an enabled state.
-		if slices.Contains(cfg.ClusterDomains, domain) {
-			state = "enabled"
-		} else {
-			state = "disabled"
-		}
+func (h *configHandler) handleIngress(req router.Request, resp router.Response, domain string, acornDNS *string, ingressClassName *string) error {
+	if acornDNS == nil || *acornDNS == "disabled" {
```

Comment:
This presents a sublte issue...if i run acorn locally, acornDNS will equal `"auto"`, so we will get past this and create the ingress and service.

But in this case, acorn will be configured to use the localhost dns entry - `local.oss-acorn.io`. That will be the domain that is set in config.ClusterDomains.

Later, in ingress/dns.go, the handler will ultimately do the right thing and NOT create the dns entries because the check it does is `if slices.Contains(cfg.ClusterDomains, domain) {` ([see here](https://github.com/acorn-io/runtime/blob/main/pkg/controller/ingress/dns.go#L65)). So, that's good - my main worry was that we were going to create dns entries when the cluster is actually configured to use local. The secondary concern is that in this case we are creating an ingress and service when we don't need to.

So, it would better to decide whether to proceed or not with the same check: `if slices.Contains(cfg.ClusterDomains, domain`, but here's where it gets tricky: the value of cfg.ClusterDomains is populated by logic that is dependent on the acornDNS secret just created - https://github.com/acorn-io/runtime/blob/main/pkg/config/config.go#L150-L176. So, the first time through, when we _actually_ create the dns secret, the config we have may be stale. And even if we got the config here again, getting it uses a cache to retrieve the secret, so it might not be caught up with the secret we just created.

I'm still thinking about how to address this, just wanted to get the problem out there.

Reply:
here's an idea: make the handleIngress a handler for the acornDNS secret. So, any time that secret changes, it would trigger the handler. Then, get the config and it should be guaranteed to be "up to date" WRT the acornDNS secret because it otherwise would not have triggered the handler. Then you'd be safe to look at cfg.ClusterDomains

Reply:
I like it, I'll get that in.